File: new/cpp/141_linked_list_cycle_hash_set.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        // idea: hash set
        // if the current node, while itereating, already exists in the hashset, return true, else return false
        unordered_set<ListNode*> seen; // set of ListNode pointers, pointing to a node's address
        ListNode *curr = head;

        while (curr) {
            if (seen.find(curr) != seen.end()) {
                // then we found it, the iterator is not at the end (if at the end, the curr is not in the hashset)
                return true;
            }
            seen.insert(curr);
            curr = curr->next;
        }

        return false;
    }
    // time: O(n)
    // space: O(n)
};

File: new/python/141_linked_list_cycle_hash_set.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # idea: using hashset
        # use hashset to check if the current node's address exists in the hashset, if it exists, return true
        # if not return false.
        seen = set()
        curr = head
        while curr is not None:
            if curr in seen:
                return True
            else:
                seen.add(curr)
                curr = curr.next

        return False
        # time: O(n)
        # space: O(n)
